samples/hello-cube.html

```diff
@@ -0,0 +1,313 @@
+        size: vertexArray.byteLength,
+        usage: GPUBufferUsage.VERTEX | GPUBufferUsage.TRANSFER_DST
+    };
+    verticesBuffer = device.createBuffer(verticesBufferDescriptor);
+    verticesBuffer.setSubData(0, vertexArray.buffer);
+
+    /* Input State */
+    const positionAttributeDescriptor = {
+        attributeIndex: 0,
+        offset: 0,
+        format: "float4"
+    };
+    const colorAttributeDescriptor = {
+        attributeIndex: 1,
+        offset: colorOffset,
+        format: "float4"
+    }
+    const vertexBufferDescriptor = {
+        stride: vertexSize,
+        stepMode: "vertex",
+        attributes: [positionAttributeDescriptor, colorAttributeDescriptor]
+    };
+    const inputStateDescriptor = {
+        indexFormat: "uint32",
+        vertexBuffers: { 0: vertexBufferDescriptor }
+    };
+
+    /* Bind Group Layout */
+    const transformBufferBindGroupLayoutBinding = {
+        binding: 0,
+        visibility: GPUShaderStageBit.VERTEX,
+        type: "uniform-buffer"
+    };
+
+    const uniformsBindGroupLayoutDescriptor = {
+        bindings: [transformBufferBindGroupLayoutBinding]
+    };
+    uniformsBindGroupLayout = device.createBindGroupLayout(uniformsBindGroupLayoutDescriptor);
+
+    const depthStateDescriptor = {
+        depthWriteEnabled: true,
+        depthCompare: "less"
+    };
+
+    /* Render Pipeline */
+    const vertexStageDescriptor = {
+        module: shaderModule,
+        entryPoint: "vertex_main"
+    };
+    const fragmentStageDescriptor = {
+        module: shaderModule,
+        entryPoint: "fragment_main"
+    };
+    const pipelineDescriptor = {
+        bindGroupLayouts: [uniformsBindGroupLayout],
+
+        vertexStage: vertexStageDescriptor,
+        fragmentStage: fragmentStageDescriptor,
+
+        primitiveTopology: "triangle-list",
+        depthStencilState: depthStateDescriptor,
+        inputState: inputStateDescriptor
+    };
+    pipeline = device.createRenderPipeline(pipelineDescriptor);
+
+    /* Renderpass Attachments */
+    let colorAttachment = {
+        // attachment is acquired in render loop.
+        loadOp: "clear",
+        storeOp: "store",
+        clearColor: { r: 0.5, g: 1.0, b: 1.0, a: 1.0 }
+    };
+
+    // GPUExtent3D
+    const depthSize = {
+        width: canvas.width,
+        height: canvas.height,
+        depth: 1
+    };
+
+    const depthTextureDescriptor = {
+        size: depthSize,
+        arrayLayerCount: 1,
+        mipLevelCount: 1,
+        sampleCount: 1,
+        dimension: "2d",
+        format: "d32-float-s8-uint",
+        usage: GPUTextureUsage.OUTPUT_ATTACHMENT
+    };
+
+    const depthTexture = device.createTexture(depthTextureDescriptor);
+
+    // GPURenderPassDepthStencilAttachmentDescriptor
+    const depthAttachment = {
+        attachment: depthTexture,
+
+        depthLoadOp = "clear",
+        depthStoreOp = "store",
+        clearDepth: 1.0
+    };
+
+    renderPassDescriptor = {
```

Comment:
Instead of writing into globals, it's probably better style to return a collection of all the variables.

Reply:
That sounds very functional indeed.